Feature/new subcommand

Aline-9 opened the pull request:
Changes:
- new structure in class PostmanCommandLineOptions --> class CommandLineParser not necessary anymore
- some changes in other classes (QbicDataStatus, QbicDataDownloader, App) due to the new structure
- format usage help message

Open:
- update README
- Remove/work on currently not used options
- some picocli functions don't work as expected (mutually exclusive options, parse over all arguments)
---> current solution for these without use of picocli

Aline-9 committed:
Added subcommands for download and status

Aline-9 committed:
change of picocli structure

KochTobi commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
+      System.out.println(
+          "You have to provide one ID as command line argument or a file containing IDs.");
+      System.exit(1);
+    } else if ((ids != null) && (filePath != null)) {
```
```suggestion
    } else if (nonNull(ids) && nonNull(filePath)) {
```
from `import static java.util.Objects.nonNull;`
As of java 8, the `java.utils.Objects` class provides some helpful methods to deal with objects. IMHO this is more readable.

Aline-9 replied:
Thank you for this tipp, I haven't heard of this before but it really is easier to read. I'll change it:)

KochTobi commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
-  public String dss_url = "https://qbis.qbic.uni-tuebingen.de/datastore_server";
+  private List<String> verifyProvidedIdentifiers() throws IOException {
+    if ((ids == null || ids.isEmpty()) && filePath == null) {
+      System.out.println(
```
Do you want to print it to System.out or System.err?

Aline-9 replied:
System.err is probably better here because it is an invalid input and therefore no normal output but an error... thanks for addressing that:)

KochTobi commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
+          "You have to provide one ID as command line argument or a file containing IDs.");
+      System.exit(1);
+    } else if ((ids != null) && (filePath != null)) {
+      System.out.println(
```
Do you want to print it to System.out or System.err?

Aline-9 replied:
same here:)

KochTobi commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
+        Authentication authentication = App.loginToOpenBIS(passwordEnvVariable, user, as_url);
+        ChecksumReporter checksumWriter =
+                new FileSystemWriter(
+                        Paths.get(
+                                System.getProperty("user.dir") + File.separator + "logs/summary_valid_files.txt"),
+                        Paths.get(
+                                System.getProperty("user.dir") + File.separator
+                                        + "logs/summary_invalid_files.txt"));
```
```suggestion
                        Paths.get(
                                System.getProperty("user.dir") + File.separator + "logs" + File.separator + "summary_valid_files.txt"),
                        Paths.get(
                                System.getProperty("user.dir") + File.separator
                                        + "logs" + File.separator + "summary_invalid_files.txt"));
```

Aline-9 replied:
👍

KochTobi commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
+      System.out.println(
+          "Arguments --identifier and --file are mutually exclusive, please provide only one.");
+      System.exit(1);
+    } else if (filePath != null) {
+      ids = IdentifierParser.readProvidedIdentifiers(filePath.toFile());
+    }
+    return ids;
```
Here you specify, that the file has precedence over the ids. Is this related to the issue you mentioned?

Aline-9 replied:
I don't think that the file has precedence over the ids. They are mutually exclusive. The first if case in this method adresses the case, that no ids or file is given. The second one is for the case, that both is given. This is not possible and there will be an error message. Therefore, the third if case handles the case that only a file is given. Because we need the ids in the file to proceed, the readProvidedIdentifiers functions is called here. The result is set to a String list ids, but this list did not exist in this case before and is therefore not overwritten.

The issue I mentioned only occures when I tried to handle this "mutually exclusive options case" with the option @ArgGroup that is provided by picocli. It should be possible to define groups that are mutually exclusive. But because it did not work like this and I also found a similar issue on the github page of picocli which also did not have a solution yet, I kept the functionality that handled the issue before in the CommandLineParser class in the function verifyProvidedIdentifiers().

KochTobi commented on src/main/java/life/qbic/model/download/Authentication.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public void login() throws ConnectionException, AuthenticationException {
+        try {
+            sessionToken = applicationServer.login(user, password);
+        } catch (Exception e) {
```
Do we need to logout somewhere? I could not find anything in the old code so I assume we don't have to :shrug:

Aline-9 replied:
I can do some research about that:)

Aline-9 replied:
Until now, we did not logout. But I found a logout method for the application server and added it at the end of the whole process... or is that unnecessary?

KochTobi commented on src/main/java/life/qbic/model/download/Authentication.java:
```diff
@@ -0,0 +1,62 @@
+import ch.ethz.sis.openbis.generic.asapi.v3.IApplicationServerApi;
+import ch.systemsx.cisd.common.spring.HttpInvokerUtils;
+
+public class Authentication {
```
Very nice abstraction :+1:

KochTobi commented on src/main/java/life/qbic/model/download/QbicDataDownloader.java:
```diff
@@ -61,23 +52,23 @@ public class QbicDataDownloader {
    *
    * @param AppServerUri  The openBIS application server URL (AS)
    * @param DataServerUri The openBIS datastore server URL (DSS)
-   * @param user          The openBIS user
-   * @param password      The openBis password
+   //* @param user          The openBIS user
+   //* @param password      The openBis password
    * @param bufferSize    The buffer size for the InputStream reader
```
```suggestion
   * @param AppServerUri  The openBIS application server URL (AS)
   * @param DataServerUri The openBIS datastore server URL (DSS)
   * @param bufferSize    The buffer size for the InputStream reader
   * @param filterType //I think you can remove this one
   * @param conservePaths // TODO document
   * @param checksumReporter // TODO document
   * @param sessionToken The session token for the openBisApp (or is it the data server?)
   ```

Aline-9 replied:
Oh right I forgot to edit this. 

But I have two questions considering the options:
- do you know for what we need the conserve Paths option? Because I used it and did not notice any change... it is used in the code but I can't really tell what the goal here is
- The filter and the regex option are currently not used and the suffix one only for the download... I think it would not be that much work to make them work again but I was wondering if there is still a need for them.

KochTobi reviewed (commented):
Hi aline,
Overall looks good :) I left some comments. I'll try to run it later and probably will leave some more comments. Feel free to address them.

Aline-9 commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
       required = true,
-      description = "openBIS user name")
+      description = "openBIS user name",
+      scope = CommandLine.ScopeType.INHERIT)
```
If the scope of an option is set to INHERIT, the option is also usable in the subcommands of the command. In our case, every option that is specified in the main command can be used in both subcommands and therefore needs this scope.

Aline-9 commented on src/main/java/life/qbic/io/commandline/PostmanCommandLineOptions.java:
```diff
@@ -1,85 +1,161 @@
+
 public class PostmanCommandLineOptions {
 
+  @Command(name = "download",
```
I thought about that, especially because the download subcommand is quite big and also has its own options. I just wanted to make it work first and also have some doubts about putting the commands in separate classes, because they share a lot of options that have to be inherited and also because i was not sure if the @Command structure of picocli supports using separate classes. All examples I saw only had one class.

Aline-9 committed:
adding logout method and other smaller changes

Aline-9 committed:
small change I forgot to add

Aline-9 committed:
- update Readme
- remove unused options
- remove logout function again

Aline-9 committed:
fixing the suffix option for the case that there is no file available for the given suffix

KochTobi committed:
Avoid exception swallowing

KochTobi committed:
Correctly filter download.

Correctly filters the downloaded files. Skips identifiers for which no files are present for the given filter rule. This needed to be done in order to avoid an influx of "Nothing to download" messages being presented to the user.

KochTobi reviewed (approved):
Very nice work LGTM :+1:
